code/centauri/algorithms/grandpa/verifier/src/lib.rs

```diff
@@ -51,57 +54,78 @@ where
 	Host: HostFunctions,
 	Host::BlakeTwo256: Hasher<Out = H256>,
 {
-	let ParachainHeadersWithFinalityProof { finality_proof, mut parachain_headers } = proof;
+	let ParachainHeadersWithFinalityProof { finality_proof, parachain_headers } = proof;
 
-	// 1. first check that target is in proof.unknown_headers.
+	// 1. First validate unknown headers.
 	let headers = AncestryChain::<H>::new(&finality_proof.unknown_headers);
-	let target = headers
-		.header(&finality_proof.block)
-		.ok_or_else(|| anyhow!("Target header with hash: {:?} not found!", finality_proof.block))?;
+	let target = finality_proof
+		.unknown_headers
+		.iter()
+		.max_by_key(|h| *h.number())
+		.ok_or_else(|| anyhow!("Unknown headers can't be empty!"))?;
 
-	// 2. next check that there exists a route from client.latest_relay_hash to target.
-	let finalized = headers
-		.ancestry(client_state.latest_relay_hash, finality_proof.block)
+	// this is illegal
+	if target.hash() != finality_proof.block {
+		Err(anyhow!("Latest finalized block should be highest block in unknown_headers"))?;
+	}
+
+	let from = finality_proof
```

Comment:
not sure if worth it:
but perhaps it's better to loop over all unknown headers once and return the min and max by key?

Reply:
> but perhaps it's better to loop over all unknown headers once and return the min and max by key?

Well given that we're pruning unknown headers, it's not worth it.